crates/sim2h/src/lib.rs

```diff
@@ -903,369 +1145,73 @@ impl Sim2h {
-                    {
-                        Some(to_url) => to_url,
-                        None => {
-                            error!("unvalidated proxy agent {}", &req_agent_id);
-                            return;
-                        }
-                    };
-                    state.send(
-                        req_agent_id,
-                        to_url,
-                        &msg_out,
-                    );
-                });
-                Ok(())
-            }
-            _ => {
-                warn!("Ignoring unimplemented message: {:?}", message );
-                Err(format!("Message not implemented: {:?}", message).into())
-            }
-        }
-    }
 
-    fn handle_unseen_aspects(
-        &self,
-        uri: &Lib3hUri,
-        space_address: &SpaceHash,
-        agent_id: &AgentId,
-        list_data: &EntryListData,
-    ) {
-        let sim2h_handle = self.sim2h_handle.clone();
-        let uri = uri.clone();
-        let space_address = space_address.clone();
-        let agent_id = agent_id.clone();
-        let list_data = list_data.clone();
-        tokio::task::spawn(async move {
-            let disconnects = sim2h_handle.lock_state().await.build_handle_unseen_aspects(
-                uri,
-                space_address,
-                agent_id,
-                list_data,
-            );
-            sim2h_handle.disconnect(disconnects);
-        });
-    }
+                // TODO - if we have multiple options,
+                // do we want to fire off more than one?
+                let query_agent = query_agents[0].clone();
 
-    fn fetch_aspects_from_arbitrary_agent(
-        sim2h_handle: Sim2hHandle,
-        aspects_to_fetch: AspectList,
-        for_agent_id: AgentId,
-        agent_pool: Vec<AgentId>,
-        space_address: SpaceHash,
-    ) {
-        tokio::task::spawn(async move {
-            let state = sim2h_handle.lock_state().await;
-            let disconnects = state.build_aspects_from_arbitrary_agent(
-                aspects_to_fetch,
-                for_agent_id,
-                agent_pool,
-                space_address,
+                let uri = match state.lookup_joined(space_hash, &query_agent) {
+                    None => continue,
+                    Some(uri) => uri,
+                };
+
+                let wire_message =
+                    WireMessage::Lib3hToClient(Lib3hToClient::HandleFetchEntry(FetchEntryData {
+                        request_id: "".to_string(),
+                        space_address: (&**space_hash).clone(),
+                        provider_agent_id: (&*query_agent).clone(),
+                        entry_address: (&**entry_hash).clone(),
+                        aspect_address_list: Some(aspects.iter().map(|a| (&**a).clone()).collect()),
+                    }));
+
+                sim2h_handle.send((&*query_agent).clone(), (&*uri).clone(), &wire_message);
```

Comment:
I'm wondering if we shouldn't also be building this up into a multi-message.  It seems likely to me that after going through this many of the HandleFetchEntries will actually be requests to the same agent.  If so they shouldn't go out as a multiple sends.

Reply:
The aspects at least are already aggregated into a single fetch entry request.

The `get_agents_for_query` function, however, randomizes agents for querying against in an effort to distribute the workload. It feels like pre-optimization complexity to me to aggregate messages if the happen to randomly be assigned to the same agent.